### ISSUE TITLE
Contact List

### ISSUE THREAD

**Issue by Crt123** [nglviewer/ngl#782]
Dear Colleagues,

I am wondering if it is possible to list contacts between receptor and ligand using NGL.  I would like to tabulate interactions between ligand and receptor for analysis.

Here is my script:

<!DOCTYPE html>
<head>
<script src="ngl.js"></script>

</head>
<body>
<div id="viewport" style="800px; height:800px;"></div>
<script>

// Create NGL Stage object
var stage = new NGL.Stage( "viewport" );

// Handle window resizing
window.addEventListener( "resize", function( event ){
    stage.handleResize();
}, false );

stage.loadFile("complex.pdb").then(function (o) {

//
// SetLigand representation
//
var ligandRepr
    ligandRepr = o.addRepresentation('hyperball', {
      multipleBond: 'symmetric',
      colorValue: 'grey',
      sele: 'UNL'
 //     aspectRatio: 1.2,
 //     radiusScale: 2.5
     })
//
//  Set neighborRepr
//	 

// get all atoms within 5 Angstrom of ligand
var selection = new NGL.Selection( "UNL");
var radius = 4;
var atomSet = o.structure.getAtomSetWithinSelection( selection, radius );
// expand selection to complete groups
var atomSet2 = o.structure.getAtomSetWithinGroup( atomSet );

 var neighborRepr
    neighborRepr = o.addRepresentation('licorice', {
      sele: atomSet2.toSeleString(),
      aspectRatio: 1.1,
      colorValue: 'lightgrey',
      multipleBond: 'symmetric',
    })

console.log(atomSet2);
console.log(atomSet2.toSeleString());

var labelRepr
    labelRepr = o.addRepresentation('label', {
      sele: atomSet2.toSeleString(),
      color: '#333333',
      yOffset: 0.2,
      zOffset: 0.8,
      attachment: 'bottom-center',
      showBorder: true,
      borderColor: 'lightgrey',
      borderWidth: 0.25,
      disablePicking: true,
      radiusType: 'size',
      radiusSize: 1.5,
      labelType: 'residue',
      labelGrouping: 'residue'
    })

 var contactRepr
 contactRepr = o.addRepresentation('contact', {
      sele: atomSet.toSeleString(),
      radiusSize: 0.12,
      weakHydrogenBond: false,
      waterHydrogenBond: false,
      backboneHydrogenBond: true,
	  hydrophobic: true
    })
 
 
  ligandRepr.setVisibility(true)
  neighborRepr.setVisibility(true)
  labelRepr.setVisibility(true)
  
  var atomSet2x = atomSet2.toSeleString()
  
  contactRepr.setSelection(atomSet2x)
  contactRepr.setVisibility(true) 
  console.log(atomSet2x)
  
  o.addRepresentation("surface", {sele: "UNL", color: "white", surfaceType: "av", opacity: 0.5, side: "front"})
  
  o.autoView("ligand")
  })
</script>
</body>
</html>

[complex.zip](https://github.com/nglviewer/ngl/files/5128425/complex.zip)

**Comment by fredludlow** [comment]
t's not part of the public API, though that could be changed. (we could make the calculateContacts in [contact.ts](https://github.com/nglviewer/ngl/blob/master/src/chemistry/interactions/contact.ts) public?).

You could try adding this line to ngl.ts and rebuilding from source.

```typescript
export { calculateContacts } from './chemistry/interactions/contact'
```

A slight word of warning: the contact representation does a fairly good job but isn't rigorously parameterised. It'll try and make a reasonable guess about protonation state at pH 7ish but the rules it uses are very simple and it makes no attempt to fix side-chain flips or tautomer issues. If you're doing a quick tabulation to assist with viewing structures it might be good enough. If you're planning to calculate interactions for many structures and perform some analysis on them then I'd be really careful about getting the protein and ligand protonation/tautomerization etc as good as possible. (Apologies if you're already familiar with all this! If not, the intro to the Protonate3D paper has a good discussion of what to look out for: https://onlinelibrary.wiley.com/doi/full/10.1002/prot.22234). Programs like reduce, protonate3d and protoss are a good place to start for getting this stuff right.

**Comment by ppillot** [comment]
I did something similar in my libmol.org project
Here is a link to the relevant function from the source: https://github.com/ppillot/libmol/blob/0932696bc8e64d75e8558029f52a26cb726fbb03/src/utils/contacts.ts#L300

<img width="1217" alt="listing contacts" src="https://user-images.githubusercontent.com/3930632/91716817-57295080-eb5e-11ea-8a4d-6ce0a605f2b9.png">

**Comment by Crt123** [comment]
Dear Fred, many thanks for your answer, I will try your recepie. For fast
review of interactions the contact list from ngl will be enough. Yes, you
are right for detailed analysis is it neccessary to use correct protonation
& tautomeric representation of protein and ligand. It would be nice to put
the option to write out contact list in the new version of public API.

All the best

Crtomir

**Comment by Crt123** [comment]
Dear Paul, many thanks for your email. You did amazing aplication,
that is *fitting
our needs.*

*All the best Crtomir*

**Comment by fredludlow** [comment]
Closing for now, but reopen if more questions or if this should be part of public API

**Comment by samuelmurail** [comment]
Hi,

First of all great work !I
I want to list all contact (Hydrogen bonds, ...) from a contact representation. I am not familiar with js and so I am wondering is it possible to use the getContactData() function with something like the NGL@2.0.0-dev.32 - interactive/ligand-viewer from the gallery.

Here is what I tried :

```
$ contactRepr.repr.getContactData()
```

But i have an error `"t is undefined"`

So I tried :

```
contactRepr.repr.bufferList[0].picking.contacts.contactStore
```

I have access to interaction type and indexes of atoms, but I don't know how to link the indexes to the structure indexes, to get for each contact the atom names and residues involved.
Thanks. Samuel
